### DISCUSSION TITLE
Testing alongside BirdNET-Pi

### DISCUSSION THREAD

**Discussion by farski** [tphakala/birdnet-go#27] [Q&A]
Hi @tphakala 

How feasible is it to test this alongside BirdNET-Pi on the same Pi? I assume it should be fine, but I haven't seen an obvious way to set the web port, so that may be the only thing that conflicts.

I only have one Pi at the moment that I can get actual results with, but if I can get this running on there I can probably help out with some UI work.

Thanks for all your work on this

**Comment by tphakala** [comment]
Hi, yes I think you can share same RPI with BirdNET-Pi and BirdNET-Go. On BirdNET-Go web server port is configurable, it defaults to 8080 which will not conflict with BirdNET-Pi.

**Reply by farski** [reply]
Great, I'll give it a shot today and report back

**Reply by farski** [reply]
My RPi3 was not loving the tflite [binary](https://github.com/tphakala/tflite_c/releases/tag/v2.14.0), so I spent some time trying to build both the BirdNET-Go to get the latest version and tflite_c, but never got both working at the same time. I did move to a different single board computer I had hanging around, and got about 97% of the way there, but when I actually went to run `birdnet-go realtime`, I got an error along the lines of "labels.txt could not be found in zip archive". I will spend some more time on it over the weekend.

**Reply by Shellfishgene** [reply]
I have Birdnet-Pi and Birdnet-Go running on the same Pi, but it's a RPi4. I think you can run the RPi3 on 32 or 64 bit OS? The tflite binaries are only 64, maybe that's it?
I had the labels.txt error when I used a Birdnet-Go config file for the previous pre-GUI version, where the locale setting changed, maybe it's the same for you?

**Reply by tphakala** [reply]
Config file changes over versions are not handled well, or at all at the moment. Locale setting has indeed moved to birdnet node, and if you have config file created by older version app startup fails. Please backup old config file to another location and let birdnet-go to create new config file, after that you can reconfigure defaults with your specific settings and it should work.

Tflite library is indeed 64-bit only, provided file should work on RPI 3 with 64-bit OS, at least I have been using it successfully, although by current dev system is RPI 5 with latest RPI OS.

**Reply by Smo-RBR** [reply]
Can you give a hint where to configure the webserver. Birdnet-Go seems to run fine aside of birdnet-pi on an Raspberry 3b, but no webserver is running, and the debug output says also nothing about.

**Reply by Shellfishgene** [reply]
Did you compile from source? The current binaries don't have the webserver yet.

**Reply by Smo-RBR** [reply]
ah, okay. I fiddled around in the code, added the config for the webserver in config file, and nothing happened :)

**Reply by tphakala** [reply]
New binary release for RPI containing web ui is now uploaded to releases section

**Comment by unosonic** [comment]
actually testing BirdNET-Pi and birdnet-go in parallel on a raspberry Pi4 (8G RAM).
I'm listenig live with BirdNET-PI, birdnet-go does realtime analysis (-t 0.1 in order to increase biodiversity in my backyard ;)
System load is about 1.5, processing time as reported by birdnet-go is about 200 - 300ms
nice!

**Comment by farski** [comment]
Finally got a chance to spend some time on this, here's some notes mostly for myself, but may be useful for other early testers:

- Set up a fresh Bookworm Lite 64 bit SD card
- Grabbed the 20240215 snapshot
- Installed `pulseaudio` via apt (had to reboot to get that working) and set the pulse daemon config
- Set lat/long in the BirdNET config file
- `birdnet-go realtime` was giving me a bunch of SoX not found errors trying to generate spectrograms, so I also did `apt install sox`, but now I have `SoX command failed` errors instead

I also tried to make the following systemd service

```ini
[Unit]
Description=BirdNET-Go realtime analysis
After=network.target
StartLimitIntervalSec=0

[Service]
Type=simple
Restart=always
RestartSec=5
User=birdnet
ExecStart=/home/birdnet/birdnet-go realtime

[Install]
WantedBy=multi-user.target
```

I very rarely use systemd, so I'm pretty bad at it, but not matter what I did trying to run birdnet-go through there resulted in `open webui.log: permission denied`

Going to let it run over the weekend, will report back.

**Reply by PovilasID** [reply]
Any news on comparison? Performance wise?

**Comment by DMontgomery40** [comment]
First of all, awesome work here, so great to see an actively developed - and improving - instance for BirdNET. 

Quick question on BirdNET-Pi / birdnet-go dual setup:  how are guys configuring pulse or sox or whatever to send the mic to both? I'm fine on http/UI end, that was easy, can't figure the sound piece.

**Reply by tphakala** [reply]
Latest version is using ALSA only, I disabled Pulse Audio support to prevent inconsistent behaviour if BirdNET-Go is run as interactive user session or service. So, if you start BirdNET-Go in realtime mode it taps into Linux kernel ALSA API so there is nothing to configure for Pulse Audio or SOX.

Thanks to @isZumpo's great contributions BirdNET-Go is now available as Docker container and I think this will be preferred deployment mode for going forward, however I need to work on some improvement to make configuring more user friendly.

Container packages are available here https://github.com/tphakala/birdnet-go/pkgs/container/birdnet-go

**Reply by tphakala** [reply]
There indeed seems to be issues with sharing sound device between BirdNET-Pi services and BirdNET-Go, problem @farski reported is this exact issue. I will try to solve this.

**Reply by unosonic** [reply]
it would be really helpful if one could choose a specific ALSA source, in case you have several input devices present.
so far I have configured a virtual alsa pcm device like this in /etc/asound.conf
```
pcm.birdmic {
    type dsnoop
    ipc_key 1234
    ipc_key_add_uid false
    ipc_perm 0666
    slave {
        pcm "hw:MYHARDWARESOUNDCARDID"   # your actual device
        channels 2
        rate 48000
        period_size 1024
        buffer_size 4096
    }
}
pcm.!default
{
    type asym
    playback.pcm
    {
        type plug
        slave.pcm "dmix"
    }
    capture.pcm
    {
        type plug
        slave.pcm "birdmic"
    }
}

```
that way I can choose "birdmic" in birdnet-pi and it also works in birdnet-go, since I've made it the default ALSA source.

**Reply by tphakala** [reply]
@DMontgomery40 I did a lot of testing today and found root cause of issues, please see my reply at https://github.com/tphakala/birdnet-go/issues/37.

@unosonic I agree, I need to study how ALSA source can be set in malgo (miniaudio go library).

**Reply by DMontgomery40** [reply]
Awesome thanks!!  I was actually just going to update that this fixed it for me, and then just saw your solution, the .asoundrc is a bit more elegant but now I'm scared to tough anything since it's working

in the screenshot below, that is the dsnoop device, had no idea i could just write "dsnoop" !
![image](https://github.com/tphakala/birdnet-go/assets/130489651/79600861-8e69-4ea0-98fb-824b8473f4fe)

**Reply by unosonic** [reply]
I've had no success in using dsnoop device directly, it did not work for shared mic access, no idea why. But I'm using this custom dsnoop as described above successfully on several raspberry pi zeros, also together with pulseaudio. /etc/pulse/default.pa:
` load-module module-alsa-source device=birdmic`
